x/rollapp/keeper/fraud_handler.go

```diff
@@ -0,0 +1,121 @@
+package keeper
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	clienttypes "github.com/cosmos/ibc-go/v6/modules/core/02-client/types"
+	tmtypes "github.com/cosmos/ibc-go/v6/modules/light-clients/07-tendermint/types"
+
+	common "github.com/dymensionxyz/dymension/v3/x/common/types"
+	"github.com/dymensionxyz/dymension/v3/x/rollapp/types"
+)
+
+// HandleFraud handles the fraud evidence submitted by the user.
+func (k Keeper) HandleFraud(ctx sdk.Context, rollappID, clientId string, fraudHeight uint64, seqAddr string) error {
+	// Get the rollapp from the store
+	rollapp, found := k.GetRollapp(ctx, rollappID)
+	if !found {
+		return sdkerrors.Wrapf(types.ErrInvalidRollappID, "rollapp with ID %s not found", rollappID)
+	}
+
+	stateInfo, err := k.FindStateInfoByHeight(ctx, rollappID, fraudHeight)
+	if err != nil {
+		return err
+	}
+
+	//check height is not finalized
+	if stateInfo.Status == common.Status_FINALIZED {
+		return sdkerrors.Wrapf(types.ErrDisputeAlreadyFinalized, "state info for height %d is already finalized", fraudHeight)
+	}
+
+	//check height is not reverted
+	if stateInfo.Status == common.Status_REVERTED {
+		return sdkerrors.Wrapf(types.ErrDisputeAlreadyReverted, "state info for height %d is already reverted", fraudHeight)
+	}
+
+	//check the sequencer for this height is the same as the one in the fraud evidence
+	if stateInfo.Sequencer != seqAddr {
+		return sdkerrors.Wrapf(types.ErrWrongProposerAddr, "sequencer address %s does not match the one in the state info", seqAddr)
+	}
+
+	// slash the sequencer, clean delayed packets
+	err = k.hooks.FraudSubmitted(ctx, rollappID, fraudHeight, seqAddr)
+	if err != nil {
+		return err
+	}
+
+	//mark the rollapp as frozen. revert all pending states to finalized
+	rollapp.Frozen = true
+	k.SetRollapp(ctx, rollapp)
+
+	lastFinalizedIdx, _ := k.GetLatestFinalizedStateIndex(ctx, rollappID)
+	finalizedStateInfo, _ := k.GetStateInfo(ctx, rollappID, lastFinalizedIdx.Index)
+
+	lastFinalizedHeight := finalizedStateInfo.StartHeight + finalizedStateInfo.NumBlocks
```

Comment:
1. I believe this is not the height you want. this is the height on the rollapp. This height can be much bigger than the currrent block height.  Ideally I think what you need here to get the latest finalized height on the hub (not on the rollapp) which can be taken by `currentHubHeight - disputePeriodInBlocks`. 
1. Also I believe the calc is wrong here. -1 is missing (if you have `StartHeight` 1 and `NumBlocks` 1 , the latest finalized height is 1 not 2)